src/piadm.sh

```diff
@@ -486,20 +488,50 @@ update_boot_sectors() {
 
 	some=0
 	for a in "${boot_devices[@]}"; do
-		# Plow through devices, even if some fail.  installboot also
-		# does loader-into-EFI-System-Partition this way.  Trailing /
-		# is important in the -b argument because boot is actually a
-		# symlink.
-		installboot -m -b /${bootfs}/boot/ /${bootfs}/boot/pmbr \
-			/${bootfs}/boot/gptzfsboot \
-			/dev/rdsk/${a}${suffix} > /dev/null 2>&1 || \
-			eecho "WARNING: Can't installboot on ${a}${suffix}"
+		if [[ "$flag" == "-d" ]]; then
+			if [[ "$suffix" == "s0" ]]; then
+				# BIOS boot, we don't care.
+				continue
+			fi
+			# otherwise mount the ESP and trash it.
+			if [[ "$tdir" == "" ]]; then
+				tdir=`mktemp -d`
+			fi
+			mount -F pcfs /dev/dsk/${a}s0 ${tdir}
+			if [[ $? -ne 0 ]]; then
+				eecho "disk $a has no PCFS ESP, it seems"
+				continue
+			fi
+			# Just take out the EFI directory, in case someone
+			# is using it for something ELSE also.
+			/bin/rm -rf ${tdir}/EFI
+			umount ${tdir}
+			# If we make it here, at least some disks had
+			# ESP and we managed to clean them out.  "some" below
+			# will get set.
+		else
+			# Plow through devices, even if some fail.
+			# installboot also does
+			# loader-into-EFI-System-Partition this way.
+			# Trailing / is important in the -b argument
+			# because boot is actually a symlink.
+			installboot -m -b /${bootfs}/boot/ \
+				/${bootfs}/boot/pmbr \
+				/${bootfs}/boot/gptzfsboot \
+				/dev/rdsk/${a}${suffix} > /dev/null 2>&1 || \
+				eecho \
+				"WARNING: Can't installboot on ${a}${suffix}"
+		fi
+
 		if [[ $? -eq 0 ]]; then
 			some=1
 		fi
```

Comment:
This whole pattern seems error prone to me. `$?` will only be `0` if the *last* `umount` succeeded. What happens if all `umount` calls succeed except the last one? It doesn't seem like it's going to do the right thing.

Reply:
Oh damn... I need to change that check to `if [[ $some -eq 0 && $? -eq 0 ]]`.  Thanks.

Reply:
Hmmmm, upon further review, if any umount calls succeed, "some" gets set to 1.  If they fail, nothing happens ("some" remains the same).  It literally indicates at least one of the disks in the pool was successfully hit.  If it's one pool and it fails, that works.

Reply:
You know what, I missed that this is happening inside the loop, and not outside. It's fine as-is.

Sorry about that :-/